src/qibolab/_core/instruments/keysight/pulse.py

```diff
@@ -0,0 +1,169 @@
+"""Utils for pulse handling."""
+
+from collections import defaultdict
+from collections.abc import Iterable
+from typing import Union
+
+from keysight import qcs
+
+from qibolab._core.pulses import Drag, Envelope, Gaussian, PulseId, Rectangular
+from qibolab._core.pulses.pulse import PulseLike
+
+NS_TO_S = 1e-9
+
+
+def generate_qcs_envelope(shape: Envelope) -> qcs.Envelope:
+    """Converts a Qibolab pulse envelope to a QCS Envelope object."""
+    if isinstance(shape, Rectangular):
+        return qcs.ConstantEnvelope()
+
+    elif isinstance(shape, (Gaussian, Drag)):
+        return qcs.GaussianEnvelope(shape.rel_sigma)
```

Comment:
So the DRAG is implemented here at https://github.com/qiboteam/qibolab/pull/944/files#diff-e3c18ab9440f0e68561a15f5e9497097a405ef090fd2612f4f3a362c71c83502R126-R127
and I need the gaussian as the base envelope for the summation of the base and derivative envelope.
(https://quantumbenchmark.com/docs/qcs/api/channels.html#rj-drag)

I could implement the DRAG based on Qibolab's implementation, but I would not like to create a static envelope especially when duration sweepers can change the envelope size.

Reply:
@sorewachigauyo I understand your concern, and now that you explained I'm even fine with your solution.

Not to enforce the change request, but as a general note, we suspect that most of the APIs are actually implementing duration sweepers in the same inefficient way you would do using Qibolab's envelopes (i.e. evaluating on software all the required arrays and uploading them).
This is not happening for QM, but there is a dedicated feature linked to that, and we're now exposing it as a separate sweeper (where the envelope is not uploaded multiple times, but interpolated on the device). 
https://github.com/qiboteam/qibolab/blob/f74c334d278f9c45de84f38e0f719a25c0dc1fc5/src/qibolab/_core/sweeper.py#L25
If you suspect Keysight to be able to interpolate as well, I'd recommend you to verify it (as much as you can), and expose it through that sweeper, instead of the bare duration one (there will be a fallback at some level, but they may have different caveats, related to timing and approximation).